presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayHashCodeOperator.java

```diff
@@ -13,48 +13,45 @@
  * limitations under the License.
  */
 
-import com.facebook.presto.metadata.FunctionRegistry;
-import com.facebook.presto.metadata.SqlOperator;
+import com.facebook.presto.spi.PrestoException;
 import com.facebook.presto.spi.block.Block;
 import com.facebook.presto.spi.type.StandardTypes;
 import com.facebook.presto.spi.type.Type;
-import com.facebook.presto.spi.type.TypeManager;
-import com.google.common.collect.ImmutableList;
+import com.facebook.presto.type.SqlType;
+import com.google.common.base.Throwables;
 
 import java.lang.invoke.MethodHandle;
-import java.util.Map;
 
 import static com.facebook.presto.metadata.OperatorType.HASH_CODE;
-import static com.facebook.presto.metadata.Signature.comparableTypeParameter;
+import static com.facebook.presto.spi.StandardErrorCode.INTERNAL_ERROR;
+import static com.facebook.presto.spi.type.TypeUtils.readNativeValue;
 import static com.facebook.presto.type.ArrayType.ARRAY_NULL_ELEMENT_MSG;
 import static com.facebook.presto.type.TypeUtils.checkElementNotNull;
-import static com.facebook.presto.util.Reflection.methodHandle;
 
+@ScalarOperator(HASH_CODE)
 public class ArrayHashCodeOperator
-        extends SqlOperator
 {
-    public static final ArrayHashCodeOperator ARRAY_HASH_CODE = new ArrayHashCodeOperator();
-    public static final MethodHandle METHOD_HANDLE = methodHandle(ArrayHashCodeOperator.class, "hash", Type.class, Block.class);
+    private ArrayHashCodeOperator() {}
 
-    private ArrayHashCodeOperator()
-    {
-        super(HASH_CODE, ImmutableList.of(comparableTypeParameter("T")), StandardTypes.BIGINT, ImmutableList.of("array(T)"));
-    }
-
-    @Override
-    public ScalarFunctionImplementation specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
-    {
-        Type type = types.get("T");
-        return new ScalarFunctionImplementation(false, ImmutableList.of(false), METHOD_HANDLE.bindTo(type), isDeterministic());
-    }
-
-    public static long hash(Type type, Block block)
+    @TypeParameter("T")
+    @SqlType(StandardTypes.BIGINT)
+    public static long hash(
+            @OperatorDependency(operator = HASH_CODE, returnType = StandardTypes.BIGINT, argumentTypes = {"T"}) MethodHandle hashFunction,
+            @TypeParameter("T") Type type,
+            @SqlType("array(T)") Block block)
     {
         int hash = 0;
         for (int i = 0; i < block.getPositionCount(); i++) {
             checkElementNotNull(block.isNull(i), ARRAY_NULL_ELEMENT_MSG);
-            // TODO: This should be migrated away from Type.hash and invoke a MethodHandle for the hash code operator instead
-            hash = (int) CombineHashFunction.getHash(hash, type.hash(block, i));
+            try {
+                hash = (int) CombineHashFunction.getHash(hash, (Long) hashFunction.invoke(readNativeValue(type, block, i)));
```

Comment:
It might be possible to call `(long) hashFunction.invokeExact(...)`to avoid boxing.